contract/models/contract.py

```diff
@@ -106,26 +107,22 @@ class ContractContract(models.Model):
         track_visibility="onchange",
     )
 
-    @api.multi
     def _inverse_partner_id(self):
         for rec in self:
             if not rec.invoice_partner_id:
                 rec.invoice_partner_id = rec.partner_id.address_get(["invoice"])[
                     "invoice"
                 ]
 
-    @api.multi
     def _get_related_invoices(self):
         self.ensure_one()
 
         invoices = (
-            self.env["account.invoice.line"]
+            self.env["account.move.line"]
             .search([("contract_line_id", "in", self.contract_line_ids.ids,)])
-            .mapped("invoice_id")
-        )
-        invoices |= self.env["account.invoice"].search(
-            [("old_contract_id", "=", self.id)]
+            .mapped("move_id")
         )
+        invoices |= self.env["account.move"].search([("old_contract_id", "=", self.id)])
```

Comment:
It's not exactly a migration only field, but a caveat we have from prior implementation (pre-v12): the only link between the contract and the invoice was at header level. Now it's contract line to invoice line. As you can't deduce each link for lines before this new implementation (invoices can be manipulated after the contract generation, etc), the commitment solution was to keep this field "forever" and search through it. For >=v12, this field will never be filled, and for <v12, it will contain data and you can't remove the field if you want that the smart-button continues showing the same number of invoices.

I have added a comment in this line clarifying its usage.